### PR TITLE
[Spark] Support OPTIMIZE tbl FULL for clustered table

### PR DESCRIPTION
**Pull request by dabao521**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

1. Add new sql syntax OPTIMIZE tbl FULL
2. Implemented OPTIMIZE tbl FULL to re-cluster all data in the table.

## How was this patch tested?

new unit tests added

## Does this PR introduce _any_ user-facing changes?

Yes
Previously clustered table won't re-cluster data that was clustered against different cluster keys. With OPTIMIZE tbl FULL, they will be re-clustered against the new keys.

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala [hunk 1]
```diff
@@ -153,7 +153,7 @@ class IncrementalZCubeClusteringSuite extends QueryTest
                 inputZCubeFiles = ClusteringFileStats(2, SKIP_CHECK_SIZE_VALUE),
                 inputOtherFiles = ClusteringFileStats(4, SKIP_CHECK_SIZE_VALUE),
                 inputNumZCubes = 1,
-                mergedFiles = ClusteringFileStats(6, SKIP_CHECK_SIZE_VALUE),
+                mergedFiles = ClusteringFileStats(4, SKIP_CHECK_SIZE_VALUE),
```
Comment (ami7o) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala:156]: Could you explain this change?

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala [hunk 2]
```diff
@@ -73,17 +73,17 @@ class IncrementalZCubeClusteringSuite extends QueryTest
       actualMetrics: ClusteringStats, expectedMetrics: ClusteringStats): Unit = {
     var finalActualMetrics = actualMetrics
     if (expectedMetrics.inputZCubeFiles.size == SKIP_CHECK_SIZE_VALUE) {
-      val stats = expectedMetrics.inputZCubeFiles
+      val stats = finalActualMetrics.inputZCubeFiles
```
Comment (ami7o) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala:76]: why is this needed?

**Review by ami7o — COMMENTED** [review]
Overall LGTM. some questions

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala [hunk 3]
```diff
@@ -281,5 +283,159 @@ class IncrementalZCubeClusteringSuite extends QueryTest
       }
     }
   }
+
+  test("OPTIMIZE FULL") {
```
Comment (ami7o) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala:287]: can we add a test case for different clusteringProvider with OPTIMIZE FULL?

File: spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala [hunk 4]
```diff
@@ -368,7 +368,8 @@ class DeltaSqlAstBuilder extends DeltaSqlBaseBaseVisitor[AnyRef] {
     OptimizeTableCommand(
       Option(ctx.path).map(string),
       Option(ctx.table).map(visitTableIdentifier),
-      Option(ctx.partitionPredicate).map(extractRawText(_)).toSeq)(interleaveBy)
+      Option(ctx.partitionPredicate).map(extractRawText(_)).toSeq,
+      DeltaOptimizeContext(isFull = ctx.FULL != null))(interleaveBy)
```
Comment (rahulsmahadev) [inline spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:372]: nit: Indent the tab space

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala (hunk 3 above)
Comment (dabao521) [reply spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala]: Added a new test `OPTIMIZE FULL - change clustering provider`

File: spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala (hunk 4 above)
Comment (rahulsmahadev) [reply spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:372]: oh I meant DeltaOptimizeContext is part of the above line so we can indent it by a tab space

Comment (rahulsmahadev) [inline spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:372]:
```suggestion
          DeltaOptimizeContext(isFull = ctx.FULL != null))(interleaveBy)
```

Comment (dabao521) [reply spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:372]: @rahulsmahadev , I think the current indention is correct. `DeltaOptimizeContext(isFull = ctx.FULL != null) ` is not part of the above line and it is its own a new parameter as previous lines.

Comment (dabao521) [reply spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:372]: As I mentioned in a different the comment, the current indention should be correct

Comment (rahulsmahadev) [reply spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala:372]: ah ok I misunderstood

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableStrategy.scala [hunk 5]
```diff
@@ -237,39 +238,58 @@ case class ClusteringStrategy(
...
-    ZCube.filterOutSingleZCubes(smallZCubeFiles).map { file =>
-      clusteringStatsCollector.outputStats.updateStats(file.addFile)
-      file.addFile
-    }.toSeq
+    // Note that ZCube.filterOutLargeZCubes requires clustered files have
+    // the same clustering columns, so skippedClusteredFiles are not included.
+    val smallZCubeFiles = ZCube.filterOutLargeZCubes(
+      candidateFiles.map(AddFileWithNumRecords.createFromFile), targetSize)
+
+    if (optimizeContext.isFull && skippedClusteredFiles.nonEmpty) {
```
Comment (rahulsmahadev) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableStrategy.scala:276]: for Optimize full shouldn't we consider the large zcube files as well ?

Comment (rahulsmahadev) [reply spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableStrategy.scala:276]: ah so skippedClusteredFiles will handle those

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala [hunk 6]
```diff
@@ -230,7 +230,7 @@ class IncrementalZCubeClusteringSuite extends QueryTest
                 inputZCubeFiles = ClusteringFileStats(2, SKIP_CHECK_SIZE_VALUE),
                 inputOtherFiles = ClusteringFileStats(4, SKIP_CHECK_SIZE_VALUE),
                 inputNumZCubes = 1,
-                mergedFiles = ClusteringFileStats(6, SKIP_CHECK_SIZE_VALUE),
+                mergedFiles = ClusteringFileStats(4, SKIP_CHECK_SIZE_VALUE),
```
Comment (rahulsmahadev) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala:233]: Im guessing this is also same reason as above

**Review by rahulsmahadev — APPROVED** [review]
LGTM! thanks for working on this

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala (hunk 6 above)
Comment (dabao521) [reply spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/IncrementalZCubeClusteringSuite.scala:233]: Yes, this is fixing the test bug introduced in https://github.com/delta-io/delta/pull/3793/files#r1815830851
